File: firmware/rpi_sensor_node/firebase_manager.py
import collections
import threading
import time
from dataclasses import dataclass
from typing import Optional, Tuple

from sensor_manager import SensorReading
from ml_inference import MLResult
from state_manager import DeviceState
from actuator_controller import RelayState

_OFFLINE_QUEUE_SIZE = 30


@dataclass
class _QueuedReading:
    sensor: SensorReading
    ml: MLResult
    state: DeviceState


class FirebaseManager:
    """
    Firebase Realtime Database client using the REST API + Email/Password auth.
    Mirrors FirebaseManager.h/.cpp — same paths, same offline queue.
    """

    def __init__(self, api_key: str, database_url: str,
                 email: str, password: str, device_id: str):
        self._api_key      = api_key
        self._db_url       = database_url.rstrip("/")
        self._email        = email
        self._password     = password
        self._device_id    = device_id
        self._id_token     = ""
        self._refresh_token = ""
        self._token_exp    = 0.0
        self._connected    = False
        self._queue: collections.deque = collections.deque(maxlen=_OFFLINE_QUEUE_SIZE)
        self._lock         = threading.Lock()
        self._requests     = None

    def begin(self) -> bool:
        try:
            import requests as req
            self._requests = req
        except ImportError:
            print("[FB] 'requests' not installed — Firebase disabled")
            return False

        try:
            url = (f"https://identitytoolkit.googleapis.com/v1/"
                   f"accounts:signInWithPassword?key={self._api_key}")
            resp = self._requests.post(url, json={
                "email": self._email,
                "password": self._password,
                "returnSecureToken": True,
            }, timeout=10)
            if resp.status_code != 200:
                print(f"[FB] Auth failed ({resp.status_code}): {resp.text[:200]}")
                return False
            body = resp.json()
            self._id_token      = body["idToken"]
            self._refresh_token = body["refreshToken"]
            self._token_exp     = time.monotonic() + int(body.get("expiresIn", 3600))
            self._connected     = True
            print(f"[FB] Connected — {self._db_url}")
            return True
        except Exception as e:
            print(f"[FB] begin() failed: {e}")
            return False

    # ── Public API (mirrors FirebaseManager.h) ────────────────────────────────

    def push_reading(self, s: SensorReading, ml: MLResult, state: DeviceState):
        if not self._connected:
            with self._lock:
                self._queue.append(_QueuedReading(s, ml, state))
            return
        if not self._push_one(s, ml, state):
            with self._lock:
                self._queue.append(_QueuedReading(s, ml, state))

    def send_heartbeat(self, state: DeviceState):
        self._patch(f"/devices/{self._device_id}/heartbeat", {
            "ts":     int(time.time()),
            "state":  state.value,
            "uptime": int(time.monotonic()),
        })
        self._patch(f"/devices/{self._device_id}", {"online": True})

    def flush_queue(self):
        with self._lock:
            items = list(self._queue)
            self._queue.clear()
        for item in items:
            self._push_one(item.sensor, item.ml, item.state)

    def set_online(self, online: bool):
        self._patch(f"/devices/{self._device_id}", {"online": online})

    def push_sign_in(self, user_id: str, user_name: str,
                     match_score: float, success: bool, anomaly_score: float):
        self._post(f"/signins/{self._device_id}", {
            "userId":      user_id,
            "userName":    user_name,
            "deviceId":    self._device_id,
            "matchScore":  match_score,
            "success":     success,
            "anomalyScore": anomaly_score,
            "ts":          int(time.time() * 1000),
        })

    def push_enrollment(self, user_id: str, name: str):
        self._put(f"/users/{user_id}", {
            "userId":     user_id,
            "name":       name,
            "deviceId":   self._device_id,
            "enrolledAt": int(time.time() * 1000),
            "active":     True,
        })

    def push_biometric_alert(self, user_id: str, alert_type: str, anomaly_score: float):
        self._post(f"/alerts/{self._device_id}", {
            "deviceId":    self._device_id,
            "alertType":   alert_type,
            "userId":      user_id,
            "anomalyScore": anomaly_score,
            "ts":          int(time.time() * 1000),
            "acknowledged": False,
        })

    def poll_commands(self) -> Optional[RelayState]:
        val = self._get(f"/devices/{self._device_id}/commands/relayOverride")
        if val == "ON":
            return RelayState.ON
        if val == "OFF":
            return RelayState.OFF
        return None

    def poll_enroll_command(self) -> Optional[Tuple[str, str]]:
        pending = self._get(f"/devices/{self._device_id}/commands/enroll/pending")
        if not pending:
            return None
        user_id = self._get(f"/devices/{self._device_id}/commands/enroll/userId") or ""
        name    = self._get(f"/devices/{self._device_id}/commands/enroll/name")   or ""
        if user_id:
            self._patch(f"/devices/{self._device_id}/commands/enroll", {"pending": False})
            return (user_id, name)
        return None

    def is_connected(self) -> bool:
        return self._connected

    # ── Private ───────────────────────────────────────────────────────────────

    def _push_one(self, s: SensorReading, ml: MLResult, state: DeviceState) -> bool:
        now_ms = int(time.time() * 1000)
        payload = {
            "temperatureC": s.temperature_c,
            "humidityPct":  s.humidity_pct,
            "lightRaw":     s.light_raw,
            "lightNorm":    s.light_norm,
            "riskScore":    ml.risk_score,
            "mlLabel":      ml.label,
            "state":        state.value,
            "ts":           now_ms,
        }
        ok1 = self._patch(f"/devices/{self._device_id}/latest", payload)
        ok2 = self._post(f"/readings/{self._device_id}", payload)
        return ok1 and ok2

    def _token(self) -> str:
        if time.monotonic() >= self._token_exp - 60:
            self._refresh_auth()
        return self._id_token

    def _refresh_auth(self):
        try:
            resp = self._requests.post(
                f"https://securetoken.googleapis.com/v1/token?key={self._api_key}",
                json={"grant_type": "refresh_token", "refresh_token": self._refresh_token},
                timeout=10,
            )
            body = resp.json()
            self._id_token      = body["id_token"]
            self._refresh_token = body["refresh_token"]
            self._token_exp     = time.monotonic() + int(body.get("expires_in", 3600))
        except Exception as e:
            print(f"[FB] Token refresh failed: {e}")

    def _url(self, path: str) -> str:
        return f"{self._db_url}{path}.json?auth={self._token()}"

    def _patch(self, path: str, data: dict) -> bool:
        try:
            r = self._requests.patch(self._url(path), json=data, timeout=5)
            return r.status_code == 200
        except Exception as e:
            print(f"[FB] PATCH {path}: {e}")
            return False

    def _put(self, path: str, data: dict) -> bool:
        try:
            r = self._requests.put(self._url(path), json=data, timeout=5)
            return r.status_code == 200
        except Exception as e:
            print(f"[FB] PUT {path}: {e}")
            return False

    def _post(self, path: str, data: dict) -> bool:
        try:
            r = self._requests.post(self._url(path), json=data, timeout=5)
            return r.status_code == 200
        except Exception as e:
            print(f"[FB] POST {path}: {e}")
            return False

    def _get(self, path: str):
        try:
            r = self._requests.get(self._url(path), timeout=5)
            return r.json() if r.status_code == 200 else None
        except Exception:
            return None

File: firmware/esp32_sensor_node/FirebaseManager.cpp
#include "FirebaseManager.h"
#include <Arduino.h>
#include <time.h>

FirebaseManager::FirebaseManager(const char* apiKey, const char* databaseUrl,
                                  const char* userEmail, const char* userPassword,
                                  const char* deviceId, const char* storageBucket)
    : _deviceId(deviceId), _connected(false), _authenticated(false),
      _queueHead(0), _queueCount(0)
{
    _fbConfig.api_key          = apiKey;
    _fbConfig.database_url     = databaseUrl;
    _fbAuth.user.email         = userEmail;
    _fbAuth.user.password      = userPassword;

    strlcpy(_storageBucket, storageBucket ? storageBucket : "", sizeof(_storageBucket));
    memset(_queue, 0, sizeof(_queue));
}

bool FirebaseManager::begin() {
    Serial.println("[FB] Initialising Firebase...");

    Firebase.begin(&_fbConfig, &_fbAuth);
    Firebase.reconnectWiFi(true);

    // Wait up to 10 s for authentication token
    uint32_t start = millis();
    while (!Firebase.ready() && millis() - start < 10000) {
        Serial.print(".");
        delay(300);
    }
    Serial.println();

    if (!Firebase.ready()) {
        Serial.println("[FB] Auth failed — operating in offline mode");
        return false;
    }

    _authenticated = true;
    _connected     = true;

    // Register device on first connect
    String path = _devicePath() + "/meta";
    FirebaseJson meta;
    meta.set("deviceId",  _deviceId);
    meta.set("firmware",  "2.0.0");
    meta.set("registeredAt", (uint32_t)time(nullptr));

    if (!Firebase.RTDB.updateNode(&_fbData, path.c_str(), &meta)) {
        Serial.printf("[FB] Device registration failed: %s\n",
                      _fbData.errorReason().c_str());
    } else {
        Serial.printf("[FB] Device '%s' registered\n", _deviceId);
    }

    setOnline(true);
    return true;
}

void FirebaseManager::pushReading(const SensorReading& s, const MLResult& ml,
                                   DeviceState state) {
    if (!_connected || !Firebase.ready()) {
        Serial.println("[FB] Offline — queuing reading");
        _enqueue(s, ml, state);
        return;
    }

    // Flush any queued readings first (FIFO order)
    if (_queueCount > 0) flushQueue();

    _pushOne(s, ml, state);
}

bool FirebaseManager::_pushOne(const SensorReading& s, const MLResult& ml,
                                DeviceState st) {
    FirebaseJson data;
    data.set("temperatureC",  s.temperatureC);
    data.set("humidityPct",   s.humidityPct);
    data.set("smokeRaw",      s.smokeRaw);
    data.set("smokePct",      s.smokePct);
    data.set("distanceCm",    s.distanceCm);
    data.set("riskScore",     ml.riskScore);
    data.set("mlLabel",       ml.label);
    data.set("pNormal",       ml.pNormal);
    data.set("pWarning",      ml.pWarning);
    data.set("pCritical",     ml.pCritical);
    data.set("state",         stateToString(st));
    data.set("ts",            (double)time(nullptr) * 1000.0);  // Unix ms for JS Date()

    // Update latest reading (overwrite) on /devices/{id}/latest
    String latestPath = _devicePath() + "/latest";
    if (!Firebase.RTDB.updateNode(&_fbData, latestPath.c_str(), &data)) {
        Serial.printf("[FB] Push failed: %s\n", _fbData.errorReason().c_str());
        return false;
    }

    // Append to time-series log under /readings/{id}/{pushId}
    String tsPath = _readingsPath();
    if (!Firebase.RTDB.push(&_fbData, tsPath.c_str(), &data)) {
        Serial.printf("[FB] History push failed: %s\n",
                      _fbData.errorReason().c_str());
        // Not fatal — latest was already updated
    }

    Serial.printf("[FB] Pushed: T=%.1f H=%.1f smoke=%.1f%% dist=%.1fcm risk=%.3f\n",
                  s.temperatureC, s.humidityPct, s.smokePct, s.distanceCm, ml.riskScore);
    return true;
}

bool FirebaseManager::pollCommands(RelayState& outRelay) {
    if (!_connected || !Firebase.ready()) return false;

    String path = _devicePath() + "/commands/relayOverride";
    if (!Firebase.RTDB.getString(&_fbDataRead, path.c_str())) return false;

    String val = _fbDataRead.stringData();
    if (val == "ON")  { outRelay = RelayState::ON;  return true; }
    if (val == "OFF") { outRelay = RelayState::OFF; return true; }
    return false;
}

void FirebaseManager::sendHeartbeat(DeviceState state) {
    if (!_connected || !Firebase.ready()) return;

    FirebaseJson hb;
    hb.set("ts",       (uint32_t)time(nullptr));
    hb.set("state",    stateToString(state));
    hb.set("heapFree", (uint32_t)ESP.getFreeHeap());
    hb.set("uptime",   (uint32_t)(millis() / 1000));

    String path = _devicePath() + "/heartbeat";
    Firebase.RTDB.updateNode(&_fbData, path.c_str(), &hb);
}

void FirebaseManager::setOnline(bool online) {
    if (!_connected || !Firebase.ready()) return;
    String path = _devicePath() + "/online";
    Firebase.RTDB.setBool(&_fbData, path.c_str(), online);
}

void FirebaseManager::flushQueue() {
    Serial.printf("[FB] Flushing %d queued readings\n", _queueCount);
    uint8_t flushed = 0;
    for (uint8_t i = 0; i < OFFLINE_QUEUE_SIZE && flushed < _queueCount; i++) {
        uint8_t idx = (_queueHead + i) % OFFLINE_QUEUE_SIZE;
        if (!_queue[idx].occupied) continue;
        if (_pushOne(_queue[idx].sensor, _queue[idx].ml, _queue[idx].state)) {
            _queue[idx].occupied = false;
            flushed++;
        } else {
            break;  // stop flushing if Firebase still failing
        }
    }
    _queueCount -= flushed;
    if (flushed > 0) _queueHead = (_queueHead + flushed) % OFFLINE_QUEUE_SIZE;
    Serial.printf("[FB] Flushed %d / remaining %d\n", flushed, _queueCount);
}

void FirebaseManager::_enqueue(const SensorReading& s, const MLResult& ml,
                                DeviceState st) {
    if (_queueCount >= OFFLINE_QUEUE_SIZE) {
        // Drop oldest reading (ring buffer overwrite)
        Serial.println("[FB] Queue full — dropping oldest reading");
        _queueHead = (_queueHead + 1) % OFFLINE_QUEUE_SIZE;
        _queueCount--;
    }
    uint8_t tail = (_queueHead + _queueCount) % OFFLINE_QUEUE_SIZE;
    _queue[tail] = { s, ml, st, true };
    _queueCount++;
}

// ── Biometric logging ─────────────────────────────────────────────────────────

void FirebaseManager::pushSignIn(const char* userId, const char* userName,
                                  float matchScore, bool success,
                                  float anomalyScore, const char* denyReason) {
    if (!_connected || !Firebase.ready()) {
        Serial.println("[FB] Offline — sign-in log dropped");
        return;
    }

    FirebaseJson data;
    data.set("userId",      userId);
    data.set("userName",    userName);
    data.set("deviceId",    _deviceId);
    data.set("matchScore",  matchScore);
    data.set("success",     success);
    data.set("anomalyScore", anomalyScore);
    data.set("denyReason",  denyReason ? denyReason : "none");
    data.set("ts",          (double)time(nullptr) * 1000.0);

    String path = String("/signins/") + _deviceId;
    Firebase.RTDB.push(&_fbData, path.c_str(), &data);

    Serial.printf("[FB] Sign-in logged: user=%s success=%s score=%.3f deny=%s\n",
                  userId, success ? "Y" : "N", matchScore,
                  denyReason ? denyReason : "none");
}

bool FirebaseManager::checkUserAuthorization(const char* userId) {
    if (!_connected || !Firebase.ready()) {
        // Offline — allow access so device stays operational
        Serial.println("[AUTH] Offline — granting access by default");
        return true;
    }

    char role[16]     = "staff";
    char devices[256] = "";

    // Read role
    String rolePath = String("/users/") + userId + "/role";
    if (Firebase.RTDB.getString(&_fbDataRead, rolePath.c_str())) {
        strlcpy(role, _fbDataRead.stringData().c_str(), sizeof(role));
    }
    Serial.printf("[AUTH] User '%s' role='%s'\n", userId, role);

    // Admin can open any door
    if (strcmp(role, "admin") == 0) return true;

    // Read allowedDevices (comma-separated string, e.g. "esp32_node_01,esp32_node_02")
    String devPath = String("/users/") + userId + "/allowedDevices";
    if (!Firebase.RTDB.getString(&_fbDataRead, devPath.c_str())) {
        Serial.printf("[AUTH] No allowedDevices for '%s' — denying\n", userId);
        return false;
    }
    strlcpy(devices, _fbDataRead.stringData().c_str(), sizeof(devices));
    Serial.printf("[AUTH] allowedDevices='%s' thisDevice='%s'\n", devices, _deviceId);

    // Check _deviceId appears as a full comma-delimited token
    const char* p = devices;
    size_t dlen = strlen(_deviceId);
    while ((p = strstr(p, _deviceId)) != nullptr) {
        bool startOk = (p == devices) || *(p - 1) == ',';
        bool endOk   = *(p + dlen) == '\0' || *(p + dlen) == ',';
        if (startOk && endOk) {
            Serial.printf("[AUTH] '%s' authorized for '%s'\n", userId, _deviceId);
            return true;
        }
        p++;
    }

    Serial.printf("[AUTH] '%s' NOT authorized for device '%s'\n", userId, _deviceId);
    return false;
}

void FirebaseManager::pushEnrollment(const char* userId, const char* name) {
    if (!_connected || !Firebase.ready()) return;

    FirebaseJson data;
    data.set("userId",      userId);
    data.set("name",        name);
    data.set("deviceId",    _deviceId);
    data.set("enrolledAt",  (double)time(nullptr) * 1000.0);

    String path = String("/users/") + userId;
    Firebase.RTDB.updateNode(&_fbData, path.c_str(), &data);

    Serial.printf("[FB] Enrollment logged: user=%s name=%s\n", userId, name);
}

void FirebaseManager::pushBiometricAlert(const char* userId,
                                          const char* alertType,
                                          float anomalyScore) {
    if (!_connected || !Firebase.ready()) return;

    FirebaseJson data;
    data.set("deviceId",    _deviceId);
    data.set("userId",      userId);
    data.set("alertType",   alertType);
    data.set("anomalyScore", anomalyScore);
    data.set("acknowledged", false);
    data.set("ts",          (double)time(nullptr) * 1000.0);

    String path = String("/alerts/") + _deviceId;
    Firebase.RTDB.push(&_fbData, path.c_str(), &data);

    Serial.printf("[FB] Alert logged: type=%s user=%s score=%.3f\n",
                  alertType, userId, anomalyScore);
}

bool FirebaseManager::pollEnrollCommand(char* outUserId, uint8_t userIdLen,
                                         char* outName,   uint8_t nameLen) {
    if (!_connected || !Firebase.ready()) return false;

    String pendingPath = _devicePath() + "/commands/enroll/pending";
    if (!Firebase.RTDB.getBool(&_fbDataRead, pendingPath.c_str())) return false;
    if (!_fbDataRead.boolData()) return false;

    // Read userId
    String uidPath = _devicePath() + "/commands/enroll/userId";
    if (!Firebase.RTDB.getString(&_fbDataRead, uidPath.c_str())) return false;
    strlcpy(outUserId, _fbDataRead.stringData().c_str(), userIdLen);

    // Read name
    String namePath = _devicePath() + "/commands/enroll/name";
    if (Firebase.RTDB.getString(&_fbDataRead, namePath.c_str())) {
        strlcpy(outName, _fbDataRead.stringData().c_str(), nameLen);
    } else {
        strlcpy(outName, outUserId, nameLen);  // fall back to userId as name
    }

    // Clear the pending flag (write path — use _fbData)
    Firebase.RTDB.setBool(&_fbData, pendingPath.c_str(), false);

    Serial.printf("[FB] Enroll command: userId=%s name=%s\n", outUserId, outName);
    return true;
}

// ── Firebase Storage upload ───────────────────────────────────────────────────

String FirebaseManager::uploadJpegToStorage(const uint8_t* buf, size_t len,
                                             const String& remotePath) {
    if (!_connected || !Firebase.ready() || _storageBucket[0] == '\0') return "";

    // URL-encode the path: replace '/' with '%2F'
    String encoded = remotePath;
    encoded.replace("/", "%2F");

    String url = String("https://firebasestorage.googleapis.com/v0/b/")
                 + _storageBucket
                 + "/o?uploadType=media&name=" + encoded;

    WiFiClientSecure secureClient;
    secureClient.setInsecure();  // prototype — skip cert pinning

    HTTPClient http;
    http.begin(secureClient, url);
    http.addHeader("Content-Type",   "image/jpeg");
    http.addHeader("Authorization",  String("Bearer ") + Firebase.getToken());
    http.addHeader("Content-Length", String(len));
    http.setTimeout(15000);

    int code = http.POST(const_cast<uint8_t*>(buf), len);
    http.end();

    if (code == 200 || code == 201) {
        Serial.printf("[FB] Storage upload OK: %s (%u B)\n", remotePath.c_str(), len);
        return remotePath;
    }

    Serial.printf("[FB] Storage upload failed: HTTP %d path=%s\n", code, remotePath.c_str());
    return "";
}

// ── Dashboard sign-in poll ────────────────────────────────────────────────────

bool FirebaseManager::pollLatestSignIn(double lastSeenTs, char* outName, uint8_t nameLen,
                                        bool& outSuccess, double& outTs) {
    if (!_connected || !Firebase.ready()) return false;

    String path = String("/signins/") + _deviceId;

    // orderBy("$key") works without a Firebase index — push IDs are time-ordered
    QueryFilter q;
    q.orderBy("$key").limitToLast(1);

    Serial.println("[FB] Polling latest sign-in...");

    if (!Firebase.RTDB.getJSON(&_fbDataSignIn, path.c_str(), &q)) {
        Serial.printf("[FB] pollLatestSignIn failed: %s\n",
                      _fbDataSignIn.errorReason().c_str());
        return false;
    }

    Serial.printf("[FB] Raw JSON: %s\n", _fbDataSignIn.jsonString().c_str());

    FirebaseJson     json;
    FirebaseJsonData tsData, nameData, successData;
    json.setJsonData(_fbDataSignIn.jsonString());

    size_t count = json.iteratorBegin();
    Serial.printf("[FB] Iterator count: %d\n", count);
    if (count == 0) { json.iteratorEnd(); return false; }

    String childKey, childVal;
    int    childType;
    json.iteratorGet(0, childType, childKey, childVal);
    json.iteratorEnd();

    Serial.printf("[FB] Child key: %s  val: %s\n", childKey.c_str(), childVal.c_str());

    FirebaseJson child;
    child.setJsonData(childVal);

    child.get(tsData, "ts");
    if (!tsData.success) {
        Serial.println("[FB] ts field missing");
        return false;
    }

    double ts = tsData.to<double>();
    Serial.printf("[FB] ts=%.0f  lastSeen=%.0f\n", ts, lastSeenTs);
    if (ts <= lastSeenTs) return false;

    outTs = ts;

    child.get(nameData,    "userName");
    child.get(successData, "success");

    const char* name = nameData.success ? nameData.to<const char*>() : "Unknown";
    strlcpy(outName, name, nameLen);
    outSuccess = successData.success ? successData.to<bool>() : false;

    Serial.printf("[FB] New sign-in → %s  success=%s\n",
                  outName, outSuccess ? "Y" : "N");
    return true;
}

void FirebaseManager::requestQrCode() {
    String path = _devicePath() + "/commands/qrRequest";
    FirebaseJson json;
    json.set("pending", true);
    json.set("ts",      (int)(millis()));
    if (Firebase.RTDB.setJSON(&_fbData, path.c_str(), &json)) {
        Serial.println("[FB] QR code request written");
    } else {
        Serial.printf("[FB] QR request failed: %s\n", _fbData.errorReason().c_str());
    }
}

// ── Path helpers ──────────────────────────────────────────────────────────────

String FirebaseManager::_devicePath() const {
    return String("/devices/") + _deviceId;
}

String FirebaseManager::_readingsPath() const {
    return String("/readings/") + _deviceId;
}
